File: app/schemas/contract.py
from datetime import datetime
from typing import Any, List, Optional
from pydantic import BaseModel, Field, field_validator
from enum import Enum


class ContractSortField(str, Enum):
    type_avis = "type_avis"
    statut = "statut"
    nature_contrat = "nature_contrat"
    categorie = "categorie"
    region = "region"
    date_publication = "date_publication"
    date_fermeture = "date_fermeture"

class SortOrder(str, Enum):
    asc = "asc"
    desc = "desc"

class ContractFilter(BaseModel):
    type_avis: Optional[List[str]] = None
    statut: Optional[List[str]] = None
    nature_contrat: Optional[List[str]] = None
    categorie: Optional[List[str]] = None
    region: Optional[List[str]] = None
    date_publication: Optional[str] = Field(None, max_length=100)
    date_fermeture: Optional[str] = Field(default=None, max_length=100)

class Document(BaseModel):
    titre: Optional[str] = Field(default=None, max_length=500)
    type: Optional[str] = Field(default=None, max_length=255)
    contenu: Optional[str] = None
    langue: Optional[str] = None
    dimension: Optional[str] = None
    nombre_page: Optional[str] = Field(default=None, max_length=50)

    @field_validator('contenu', 'langue', 'dimension', mode='before')
    @classmethod
    def flatten_lookup(cls, v: Any) -> Optional[str]:
        if isinstance(v, dict):
            return v.get('descriptionen') or v.get('descriptionfr') or None
        return v

class ContractBase(BaseModel):
    # ── Identity ──────────────────────────────────────────────────────────────
    numero:           Optional[str] = Field(default=None, max_length=255)
    numero_reference: str = Field(..., max_length=255)
    type_avis:        str = Field(..., max_length=100)
    statut:           str = Field(..., max_length=100)
    url:              str = Field(..., max_length=500)

    # ── Core description ──────────────────────────────────────────────────────
    titre:        str = Field(..., max_length=1000)
    organisation: str = Field(..., max_length=500)
    nature_contrat: str = Field(..., max_length=255)
    categorie:    str = Field(..., max_length=500)
    description:  Optional[str] = None

    # ── Classification ────────────────────────────────────────────────────────
    classifications: list[str] = Field(default_factory=list)
    documents:       list[Document] = Field(default_factory=list)

    # ── Geography & trade agreements ──────────────────────────────────────────
    region:      str = Field(..., max_length=500)
    accord:      Optional[str] = Field(default=None, max_length=1000)
    territoires: Optional[str] = Field(default=None, max_length=500)

    # ── Dates ─────────────────────────────────────────────────────────────────
    date_publication:           str = Field(..., max_length=100)
    date_fermeture:             Optional[str] = Field(default=None, max_length=255)
    date_limite_plaintes:       Optional[str] = Field(default=None, max_length=100)
    date_limite_interet:        Optional[str] = Field(default=None, max_length=100)
    date_ouverture_soumissions: Optional[str] = Field(default=None, max_length=100)
    date_conclusion:            Optional[str] = Field(default=None, max_length=100)

    # ── Contract terms ────────────────────────────────────────────────────────
    duree_contrat:              Optional[str] = Field(default=None, max_length=50)
    duree_contrat_avec_options: Optional[str] = Field(default=None, max_length=50)
    options_renouvellement:     Optional[str] = Field(default=None, max_length=1000)
    contrat_execution_demande:  Optional[str] = Field(default=None, max_length=50)
    contrat_a_commandes:        Optional[str] = Field(default=None, max_length=50)

    # ── Submission logistics ──────────────────────────────────────────────────
    soumission_electronique: Optional[str] = Field(default=None, max_length=255)
    endroit_reception:       Optional[str] = Field(default=None, max_length=500)
    endroit_ouverture:       Optional[str] = Field(default=None, max_length=500)
    adjudication_par_lot:    Optional[str] = Field(default=None, max_length=50)
    remarque:                Optional[str] = Field(default=None, max_length=20000)

    # ── Bid guarantee ─────────────────────────────────────────────────────────
    garantie_nature: Optional[str] = Field(default=None, max_length=500)
    garantie_valeur: Optional[str] = None

    # ── Intention-specific ────────────────────────────────────────────────────
    fournisseur_pressenti: Optional[str] = Field(default=None, max_length=500)

    # ── Contact ───────────────────────────────────────────────────────────────
    contact_nom:       Optional[str] = Field(default=None, max_length=255)
    contact_email:     Optional[str] = Field(default=None, max_length=255)
    contact_telephone: Optional[str] = Field(default=None, max_length=100)

    @field_validator('classifications', mode='before')
    @classmethod
    def parse_classifications(cls, v: Any) -> list:
        if isinstance(v, str):
            return [s.strip() for s in v.split(';') if s.strip()] if v else []
        return v if v is not None else []

class ContractResponse(ContractBase):
    id: int = Field(..., ge=0)

    model_config = {"from_attributes": True}


class ContractFilterResponse(BaseModel):
    skip: int = Field(0, ge=0)
    limit: int = Field(20, ge=1, le=100)
    total: int = Field(0, ge=0)
    contracts: Optional[List[ContractResponse]]

    model_config = {"from_attributes": True}


class ScoreBreakdown(BaseModel):
    expertise: bool
    sector: bool
    region: bool
    contract_type: bool
    sme_reserved: bool


# Deliberately not ContractResponse-based: the /recommended listing only
# loads the columns this card actually renders (see _LISTING_COLUMNS in
# app/repositories/contract.py) — a schema requiring the other ~27 Contract
# fields would force-load them right back via lazy attribute access,
# defeating the point of load_only(). Full contract detail still lives at
# GET /contract/{id} for the click-through.
class RecommendedContract(BaseModel):
    id: int = Field(..., ge=0)
    titre: str
    organisation: str
    statut: str
    nature_contrat: str
    categorie: str
    region: str
    type_avis: str
    date_publication: str
    date_fermeture: Optional[str] = None
    match_score: int = Field(..., ge=0, le=100)
    score_breakdown: ScoreBreakdown

    model_config = {"from_attributes": True}


class RecommendedContractsResponse(BaseModel):
    limit: int = Field(20, ge=1, le=100)
    total: int = Field(0, ge=0)
    capped_total: int = Field(0, ge=0)
    closing_soon_count: int = Field(0, ge=0)
    next_cursor: Optional[str] = None
    explorer_url: str
    contracts: List[RecommendedContract]

    model_config = {"from_attributes": True}


# ── Feedback (save / not relevant) ────────────────────────────────────────────
class FeedbackAction(str, Enum):
    saved = "saved"
    not_relevant = "not_relevant"


class ContractFeedbackRequest(BaseModel):
    action: FeedbackAction


class ContractFeedbackResponse(BaseModel):
    contract_id: int
    action: FeedbackAction
    created_at: datetime

    model_config = {"from_attributes": True}


class SavedContract(ContractResponse):
    saved_at: datetime


class SavedContractsResponse(BaseModel):
    skip: int = Field(0, ge=0)
    limit: int = Field(20, ge=1, le=100)
    total: int = Field(0, ge=0)
    contracts: List[SavedContract]

    model_config = {"from_attributes": True}


# ── Explorer (search + facets + keyset) ───────────────────────────────────────
class ExplorerSort(str, Enum):
    date_fermeture = "date_fermeture"      # default: most urgent (closes soonest) first
    date_publication = "date_publication"  # most recently published first
    pertinence = "pertinence"              # ts_rank desc — only meaningful with q; falls back otherwise


class ExplorerMatchMode(str, Enum):
    # Applies compatible_contracts_query(profile) server-side — the same
    # sector-OR-expertise test /recommended and the Analytics blocks qualify
    # on — instead of the frontend approximating it as a categorie= list.
    profil = "profil"


# Same rationale as RecommendedContract: only the columns the card renders,
# so load_only() on the listing query isn't immediately undone by a schema
# that force-loads the rest through lazy attribute access.
class ExplorerContract(BaseModel):
    id: int = Field(..., ge=0)
    titre: str
    organisation: str
    statut: str
    nature_contrat: str
    categorie: str
    region: str
    type_avis: str
    date_publication: str
    date_fermeture: Optional[str] = None

    model_config = {"from_attributes": True}


class FacetOption(BaseModel):
    value: str
    count: int


class ExplorerFacets(BaseModel):
    statut: List[FacetOption]
    region: List[FacetOption]
    nature_contrat: List[FacetOption]
    categorie: List[FacetOption]


class ExplorerContractsResponse(BaseModel):
    limit: int = Field(20, ge=1, le=100)
    total: int = Field(0, ge=0)
    next_cursor: Optional[str] = None
    facets: ExplorerFacets
    contracts: List[ExplorerContract]

    model_config = {"from_attributes": True}


class OrganisationSuggestion(BaseModel):
    name: str
    count: int

File: client/api/contract.ts
const API = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000/v1";

async function request<T>(path: string, options: RequestInit = {}): Promise<T> {
  const res = await fetch(`${API}${path}`, {
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
      ...options.headers,          
    },
    ...options,
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({ detail: "Request failed" }));
    throw new Error(body.detail);
  }
  return res.status === 204 ? null as T : res.json();
}

export interface Document {
  titre:       string | null;
  type:        string | null;
  contenu:     string | null;
  langue:      string | null;
  dimension:   string | null;
  nombre_page: string | null;
}

export interface Contract {
  // Identity
  id:               number;
  numero:           string;
  numero_reference: string;
  type_avis:        string;
  statut:           string;
  url:              string;

  // Core description
  titre:          string;
  organisation:   string;
  nature_contrat: string;
  categorie:      string;
  description:    string | null;

  // Classification
  classifications: string[];
  documents:       Document[];

  // Geography
  region:      string;
  accord:      string;
  territoires: string | null;

  // Dates
  date_publication:           string;
  date_fermeture:             string | null;
  date_limite_plaintes:       string | null;
  date_limite_interet:        string | null;
  date_ouverture_soumissions: string | null;
  date_conclusion:            string | null;

  // Contract terms
  duree_contrat:              string | null;
  duree_contrat_avec_options: string | null;
  options_renouvellement:     string | null;
  contrat_execution_demande:  string | null;
  contrat_a_commandes:        string | null;

  // Submission logistics
  soumission_electronique: string | null;
  endroit_reception:       string | null;
  endroit_ouverture:       string | null;
  adjudication_par_lot:    string | null;
  remarque:                string | null;

  // Bid guarantee
  garantie_nature: string | null;
  garantie_valeur: string | null;

  // Intention-specific
  fournisseur_pressenti: string | null;

  // Contact
  contact_nom:       string | null;
  contact_email:     string | null;
  contact_telephone: string | null;
}

export interface ContractFilterResponse {
  skip:      number;
  limit:     number;
  total:     number;
  contracts: Contract[] | null;
}

export interface ScoreBreakdown {
  expertise:      boolean;
  sector:         boolean;
  region:         boolean;
  contract_type:  boolean;
  sme_reserved:   boolean;
}

// Slimmer than Contract on purpose — the /recommended listing only loads (and
// only ever returns) the columns the card actually renders; full detail is a
// separate fetch at GET /contract/{id} on click-through.
export interface RecommendedContract {
  id:                number;
  titre:             string;
  organisation:      string;
  statut:            string;
  nature_contrat:    string;
  categorie:         string;
  region:            string;
  type_avis:         string;
  date_publication:  string;
  date_fermeture:    string | null;
  match_score:       number;
  score_breakdown:   ScoreBreakdown;
}

export interface RecommendedContractsResponse {
  limit:               number;
  total:               number;
  capped_total:        number;
  closing_soon_count:  number;
  next_cursor:         string | null;
  explorer_url:        string;
  contracts:           RecommendedContract[];
}

export type FeedbackAction = "saved" | "not_relevant";

export interface ContractFeedbackResponse {
  contract_id: number;
  action:      FeedbackAction;
  created_at:  string;
}

export interface SavedContract extends Contract {
  saved_at: string;
}

export interface SavedContractsResponse {
  skip:      number;
  limit:     number;
  total:     number;
  contracts: SavedContract[];
}

export const contractApi = {
  get_contracts(filter: Record<string, string | string[]>, skip: number, limit: number): Promise<ContractFilterResponse> {
    const params = new URLSearchParams()
    params.append("skip", String(skip))
    params.append("limit", String(limit))
    for (const [key, value] of Object.entries(filter)) {
      if (Array.isArray(value)) {
        value.forEach(v => params.append(key, v))
      } else {
        params.append(key, value)
      }
    }
    return request<ContractFilterResponse>(`/contract/?${params.toString()}`, { method: "GET" })
  },

  get_contract(id: number): Promise<Contract> {
    return request<Contract>(`/contract/${id}`, { method: "GET" })
  },

  get_recommended_contracts(cursor: string | null, limit: number, token: string): Promise<RecommendedContractsResponse> {
    const params = new URLSearchParams()
    if (cursor) params.append("cursor", cursor)
    params.append("limit", String(limit))
    return request<RecommendedContractsResponse>(`/contract/recommended?${params.toString()}`, {
      method: "GET",
      headers: { Authorization: `Bearer ${token}` },
    })
  },

  get_saved_contracts(skip: number, limit: number, token: string): Promise<SavedContractsResponse> {
    const params = new URLSearchParams()
    params.append("skip", String(skip))
    params.append("limit", String(limit))
    return request<SavedContractsResponse>(`/contract/saved?${params.toString()}`, {
      method: "GET",
      headers: { Authorization: `Bearer ${token}` },
    })
  },

  set_feedback(contractId: number, action: FeedbackAction, token: string): Promise<ContractFeedbackResponse> {
    return request<ContractFeedbackResponse>(`/contract/${contractId}/feedback`, {
      method: "POST",
      headers: { Authorization: `Bearer ${token}` },
      body: JSON.stringify({ action }),
    })
  },

  remove_feedback(contractId: number, token: string): Promise<null> {
    return request<null>(`/contract/${contractId}/feedback`, {
      method: "DELETE",
      headers: { Authorization: `Bearer ${token}` },
    })
  },
};